include/bredis/impl/async_op.ipp

```diff
@@ -0,0 +1,104 @@
+//
+//
+// Copyright (c) 2017 Ivan Baidakou (basiliscos) (the dot dmol at gmail dot com)
+//
+// Distributed under the MIT Software License
+//
+#pragma once
+
+#include <memory>
+#include <utility>
+
+#include <boost/asio.hpp>
+
+namespace bredis {
+
+template <typename NextLayer, typename DynamicBuffer, typename ReadCallback>
+class async_read_op {
+    NextLayer &stream_;
+    DynamicBuffer &rx_buff_;
+    ReadCallback callback_;
+    std::size_t replies_count_;
+    bool done_ = false;
+
+  public:
+    async_read_op(async_read_op &&) = default;
+    async_read_op(const async_read_op &) = default;
+
+    template <class DeducedHandler>
+    async_read_op(DeducedHandler &&deduced_handler, NextLayer &stream,
+                  DynamicBuffer &rx_buff, std::size_t replies_count)
+        : stream_(stream), rx_buff_(rx_buff), replies_count_(replies_count),
+          callback_(std::forward<ReadCallback>(deduced_handler)) {}
+
+    void operator()(boost::system::error_code, std::size_t bytes_transferred);
+
+    friend bool asio_handler_is_continuation(async_read_op *op) {
+        using boost::asio::asio_handler_is_continuation;
+        return op->done_ ||
+               asio_handler_is_continuation(std::addressof(op->callback_));
+    }
+
+    friend void *asio_handler_allocate(std::size_t size, async_read_op *op) {
+        using boost::asio::asio_handler_allocate;
+        return asio_handler_allocate(size, std::addressof(op->callback_));
+    }
+
+    friend void asio_handler_deallocate(void *p, std::size_t size,
+                                        async_read_op *op) {
+        using boost::asio::asio_handler_deallocate;
+        return asio_handler_deallocate(p, size, std::addressof(op->callback_));
+    }
+
+    template <class Function>
+    friend void asio_handler_invoke(Function &&f, async_read_op *op) {
+        using boost::asio::asio_handler_invoke;
+        return asio_handler_invoke(f, std::addressof(op->callback_));
+    }
+};
+
+template <typename NextLayer, typename DynamicBuffer, typename ReadCallback>
+void async_read_op<NextLayer, DynamicBuffer, ReadCallback>::
+operator()(boost::system::error_code error_code, std::size_t bytes_transferred) {
+    using Iterator = typename to_iterator<DynamicBuffer>::iterator_t;
+
+    done_ = true;
```

Comment:
hah nice try but you have to set this AFTER the first asynchronous call completes :)